crypto/bn/asm/rsaz-avx512.pl

```diff
@@ -0,0 +1,741 @@
+    vpsrlq    \$52, $_R2,   $T2
+
+    # "Shift left" T0..T2 by 1 QW
+    valignq \$3, $T1h,  $T2,  $T2
+    valignq \$3, $T1,   $T1h, $T1h
+    valignq \$3, $T0h,  $T1,  $T1
+    valignq \$3, $T0,   $T0h, $T0h
+    valignq \$3, $zero, $T0,  $T0
+
+    # Drop "carries" from R0..R2 QWs
+    vpandq    $mask52x4, $_R0,  $_R0
+    vpandq    $mask52x4, $_R0h, $_R0h
+    vpandq    $mask52x4, $_R1,  $_R1
+    vpandq    $mask52x4, $_R1h, $_R1h
+    vpandq    $mask52x4, $_R2,  $_R2
+
+    # Sum R0..R2 with corresponding adjusted carries
+    vpaddq  $T0,  $_R0,  $_R0
+    vpaddq  $T1,  $_R1,  $_R1
+    vpaddq  $T2,  $_R2,  $_R2
+    vpaddq  $T0h, $_R0h, $_R0h
+    vpaddq  $T1h, $_R1h, $_R1h
+
+    # Now handle carry bits from this addition
+    # Get mask of QWs whose 52-bit parts overflow...
+    vpcmpuq   \$1, $_R0,  $mask52x4, %k1 # OP=lt
+    vpcmpuq   \$1, $_R0h, $mask52x4, %k2
+    vpcmpuq   \$1, $_R1,  $mask52x4, %k3
+    vpcmpuq   \$1, $_R1h, $mask52x4, %k4
+    vpcmpuq   \$1, $_R2,  $mask52x4, %k5
+    kmovb   %k1, %r14d                   # k1
+    kmovb   %k2, %r13d                   # k1h
+    kmovb   %k3, %r12d                   # k2
+    kmovb   %k4, %r11d                   # k2h
+    kmovb   %k5, %r10d                   # k3
+
+    # ...or fully fulfilled
+    vpcmpuq   \$0, $_R0,  $mask52x4, %k1 # OP=eq
+    vpcmpuq   \$0, $_R0h, $mask52x4, %k2
+    vpcmpuq   \$0, $_R1,  $mask52x4, %k3
+    vpcmpuq   \$0, $_R1h, $mask52x4, %k4
+    vpcmpuq   \$0, $_R2,  $mask52x4, %k5
+    kmovb   %k1, %r9d                    # k4
+    kmovb   %k2, %r8d                    # k4h
+    kmovb   %k3, %ebx                    # k5
+    kmovb   %k4, %ecx                    # k5h
+    kmovb   %k5, %edx                    # k6
+
+    # Get mask of QWs where carries shall be propagated to
+    lea    (%r9d, %r14d, 2), %r14d       # k1  = k4 + 2*k1
+    xorl   %r14d, %r9d                   # k1 ^= k4
+
+    shr    \$4, %r14b
+    add    %r14b, %r8b
+    lea    (%r8d, %r13d, 2), %r13d       # k1h  = k4h + 2*k1h + k1_CF
+    xorl   %r13d, %r8d                   # k1h ^= k4h
+
+    shr    \$4, %r13b
+    add    %r13b, %bl
+    lea    (%ebx, %r12d, 2), %r12d       # k2  = k5 + 2*k2 + k1h_CF
+    xorl   %r12d, %ebx                   # k2 ^= k5
+
+    shr    \$4, %r12b
+    add    %r12b, %cl
+    lea    (%ecx, %r11d, 2), %r11d       # k2h  = k5h + 2*k2h + k2_CF
+    xorl   %r11d, %ecx                   # k2h ^= k5h
+
+    shr    \$4, %r11b
+    add    %r11b, %dl
+    lea    (%edx, %r10d, 2), %r10d       # k3  = k6 + 2*k3 + k2h_CF
+    xorl   %r10d, %edx                   # k3 ^= k6
+
+    kmovb   %r9d, %k1
+    kmovb   %r8d, %k2
+    kmovb   %ebx, %k3
+    kmovb   %ecx, %k4
+    kmovb   %edx, %k5
+
+    # Add carries according to the obtained mask
+    vpsubq  $mask52x4, $_R0,  ${_R0}{%k1}
+    vpsubq  $mask52x4, $_R1,  ${_R1}{%k2}
+    vpsubq  $mask52x4, $_R2,  ${_R2}{%k3}
+    vpsubq  $mask52x4, $_R0h, ${_R0h}{%k4}
+    vpsubq  $mask52x4, $_R1h, ${_R1h}{%k5}
```

Comment:
Is the correct %kx applied to each $_Rx here?
Looks like at this point that the order k1, k2, k3, k4, k5 aligns with the order R0, R0h, R1, R1h, R2
but here it is different, with the order being R0, R1, R2, R0h, R1h

Reply:
Fixed, thank you.